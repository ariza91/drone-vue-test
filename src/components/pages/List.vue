<template>
  <div class="list">
    <DronePageWrapper :content="wrapper">
      <template slot="title">
        <DroneListTitle class="title" />
      </template>
      <template slot="content" slot-scope="slotProps">
        <div
          class="grid"
          v-if="slotProps.show && listViewOptions.grid === getListView"
        >
          <DronePreviewGrid
            v-for="drone in droneSorted"
            :key="drone.drone_id"
            :drone="drone"
            class="drone col-12 col-sm-3"
          />
        </div>
        <div v-if="slotProps.show && listViewOptions.list === getListView">
          <DronePreviewList :drones="droneSorted" class="col-12" />
        </div>
      </template>
    </DronePageWrapper>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
import DronePageWrapper from "@/components/modules/DronePageWrapper";
import DroneListTitle from "@/components/modules/DroneListTitle";
import DronePreviewGrid from "@/components/modules/DronePreviewGrid";
import DronePreviewList from "@/components/modules/DronePreviewList";
import { compareAllElementsOf } from "@/utils/filter";
import sort from "@/utils/sort";
import listViewOptions from "@/constants/list-view";

export default {
  name: "List",
  components: {
    DronePageWrapper,
    DroneListTitle,
    DronePreviewGrid,
    DronePreviewList,
  },
  data() {
    return {
      listViewOptions,
    };
  },
  computed: {
    ...mapGetters({
      getDroneOrder: "getDroneOrder",
      getDroneList: "getDroneList",
      getListView: "getListView",
      getSearchInput: "getSearchInput",
    }),
    droneSorted() {
      if (!this.getDroneList) return [];
      const { key, order } = this.getDroneOrder;
      return this.getDroneList
        .filter(
          (drone) =>
            !this.getSearchInput ||
            compareAllElementsOf(drone, this.getSearchInput)
        )
        .sort((a, b) => sort.byNumber(a, b, key, order));
    },
    wrapper() {
      return !this.droneSorted.length ? this.$t("DroneList.notFound") : null;
    },
  },
  created() {
    this.$store.dispatch("getDroneList");
  },
};
</script>

<style scoped lang="scss">
@import "@/style/theme.scss";

.drone-list {
  height: 100%;
}

.grid {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
}
</style>