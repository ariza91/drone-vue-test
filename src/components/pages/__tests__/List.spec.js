import { createLocalVue, shallowMount } from "@vue/test-utils";
import List from "@/components/pages/List.vue";
import Vuex from 'vuex'
import orderOptions from '@/constants/drone-order'

jest.mock('@/store');
const localVue = createLocalVue()
localVue.use(Vuex)

const mockOrder = orderOptions.batteryASC
const mockDroneList = [
    { id: 1, batery: 10 },
    { id: 2, batery: 20 },
]

const factory = (store) => shallowMount(List, {
    localVue,
    store,
    mocks: { $t: (value) => value },
    stubs: ['router-link']
});

describe("List", () => {
    let cmp;
    let getters;
    let actions;
    let store;

    beforeEach(() => {
        getters = {
            getDroneOrder: () => mockOrder,
            getDroneList: jest.fn(),
            getListView: jest.fn(),
            getSearchInput: jest.fn(),
        }
        actions = {
            getDroneList: jest.fn()
        };
        store = new Vuex.Store({ getters, actions })
        cmp = factory(store);
        jest.clearAllMocks()
    })

    describe('renders', () => {
        it("renders correctly", () => {
            expect(cmp).toMatchSnapshot()
        });
    })
    describe('computed', () => {
        describe('droneSorted', () => {
            it('not droneList then empty array', () => {
                expect(cmp.vm.droneSorted).toEqual([])
            })
            it('droneList and batery ASC then array sorted', () => {
                getters = { ...getters, getDroneList: () => mockDroneList }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.droneSorted).toEqual(mockDroneList)
            })
            it('droneList and batery ASC then array sorted', () => {
                getters = {
                    ...getters,
                    getDroneList: () => mockDroneList,
                    getDroneOrder: () => orderOptions.batteryDESC
                }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.droneSorted).toEqual(mockDroneList.reverse())
            })
        })
        describe('wrapper', () => {
            it('not drone list then error', () => {
                expect(cmp.vm.wrapper).toEqual("DroneList.notFound");
            })
            it('drone list then null', () => {
                getters = { ...getters, getDroneList: () => mockDroneList }
                store = new Vuex.Store({ getters, actions })
                cmp = factory(store)
                expect(cmp.vm.wrapper).toEqual(null);
            })
        })
    })
    describe('created', () => {
        it('getDroneList called', () => {
            const spy = jest.spyOn(cmp.vm.$store, 'dispatch');
            cmp = factory(store);
            expect(spy).toHaveBeenCalledWith("getDroneList");
        })
    })
});
