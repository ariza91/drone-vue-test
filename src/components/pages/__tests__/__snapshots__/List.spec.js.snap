// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders renders correctly 1`] = `
<div class="list">
  <dronepagewrapper-stub content="DroneList.notFound"><template></template></dronepagewrapper-stub>
</div>
`;
